### PR TITLE
Stop incorrect TaskCanceledExceptions from HttpRequestPipeReader

### LINKED ISSUE #33373: Quarantine Http2StreamTests .ContentLength_Received_MultipleDataFramesOverSize_Reset
https://dev.azure.com/dnceng/public/_build/results?buildId=1176635&view=ms.vss-test-web.build-test-results-tab&runId=35414634&resultId=106018&paneView=debug

### PR DESCRIPTION
**Pull request by halter73**
Addresses #33373

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Sniped 😆

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestPipeReader.cs [hunk 1]
```diff
@@ -93,42 +93,41 @@ public void StopAcceptingReads()
 
         public void Abort(Exception? error = null)
         {
-            // We don't want to throw an ODE until the app func actually completes.
             // If the request is aborted, we throw a TaskCanceledException instead,
             // unless error is not null, in which case we throw it.
-            if (_state != HttpStreamState.Closed)
+
+            if (error is object)
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestPipeReader.cs:99]: How come not:
```suggestion
            if (error is not null)
```

**Review by JamesNK — COMMENTED** [review]
Unit test?

**Comment by halter73** [comment]
> Unit test?

This is fixing a hard-to-repro race condition reported by #33373

**Comment by JamesNK** [comment]
If it's not possible to unit test then that's fine.

**Review by JamesNK — APPROVED** [review]

**Review by BrennanConroy — APPROVED** [review]

**Review by TanayParikh — APPROVED** [review]

**Pull request merged**